File: public/python/060-unique-paths.py
# LeetCode Problem #62: Unique Paths
# Difficulty: Medium
# Link: https://leetcode.com/problems/unique-paths/

from functools import lru_cache
from math import comb

# ─────────────────────────────────────────────
# APPROACH 1: Recursive Memoization | O(mn) time | O(mn) space
# EXPLAIN: At each cell, paths = paths from cell below + paths from cell to the right.
# WHEN: Cleanest recursive derivation — base cases are the bottom row and right column.

def uniquePaths_memo(m: int, n: int) -> int:
    @lru_cache(maxsize=None)
    def dp(r, c):
        if r == m - 1 or c == n - 1:
            return 1
        return dp(r + 1, c) + dp(r, c + 1)
    return dp(0, 0)


# APPROACH 2: Bottom-Up 2D DP | O(mn) time | O(mn) space
# EXPLAIN: dp[r][c] = number of unique paths to reach (r,c) from (0,0).
# WHEN: Standard interview DP — explicit grid table is easy to trace.

def uniquePaths_dp(m: int, n: int) -> int:
    dp = [[1] * n for _ in range(m)]
    for r in range(1, m):
        for c in range(1, n):
            dp[r][c] = dp[r - 1][c] + dp[r][c - 1]
    return dp[m - 1][n - 1]


# APPROACH 3: Math (Combinatorics) | O(min(m,n)) time | O(1) space
# EXPLAIN: Total steps = (m-1) down + (n-1) right; choose which steps go down = C(m+n-2, m-1).
# WHEN: Best when you recognise the combinatorial structure — one line, no grid needed.

def uniquePaths_math(m: int, n: int) -> int:
    return comb(m + n - 2, m - 1)


# APPROACH 4: Space-Optimized DP (single row) | O(mn) time | O(n) space
# EXPLAIN: Re-use a single row; dp[j] += dp[j-1] encodes the 2D recurrence in-place.
# WHEN: Memory-efficient solution for large grids; drop-in for the 2D approach.

def uniquePaths_1d(m: int, n: int) -> int:
    dp = [1] * n
    for _ in range(1, m):
        for j in range(1, n):
            dp[j] += dp[j - 1]
    return dp[n - 1]


# APPROACH 5: Two-Row Rolling DP | O(mn) time | O(n) space
# EXPLAIN: Alternate prev/curr arrays explicitly — clearer separation of rows.
# WHEN: When you prefer explicit prev/curr over in-place single-row mutation.

def uniquePaths_two_row(m: int, n: int) -> int:
    if m == 1 or n == 1:
        return 1
    prev = [1] * n
    for _ in range(1, m):
        curr = [1] * n
        for j in range(1, n):
            curr[j] = curr[j - 1] + prev[j]
        prev = curr
    return prev[n - 1]


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [
        (3, 7, 28),
        (3, 2, 3),
        (1, 1, 1),
        (7, 3, 28),
        (3, 3, 6),
    ]
    for fn in (uniquePaths_memo, uniquePaths_dp, uniquePaths_math,
               uniquePaths_1d, uniquePaths_two_row):
        for m, n, expected in cases:
            result = fn(m, n)
            assert result == expected, (
                f'{fn.__name__}({m}, {n}) = {result}, expected {expected}'
            )
    print('All tests passed.')

# Made with Bob

File: public/java/060-unique-paths.java
/*
 * LeetCode Problem #62: Unique Paths
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/unique-paths/
 */
import java.util.*;

class Solution {

    // APPROACH 1: 2D DP | O(m*n) time | O(m*n) space
    // EXPLAIN: dp[i][j] = paths from (0,0) to (i,j); each cell = sum of cell above and cell to the left.
    // WHEN: Use for clear visualization of how paths accumulate — great for teaching.

    public int uniquePaths_2d(int m, int n) {
        int[][] dp = new int[m][n];
        for (int i = 0; i < m; i++) dp[i][0] = 1;
        for (int j = 0; j < n; j++) dp[0][j] = 1;
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[m - 1][n - 1];
    }

    // APPROACH 2: Space-Optimized DP (single row) | O(m*n) time | O(n) space
    // EXPLAIN: Re-use a single row array; dp[j] += dp[j-1] effectively computes the 2D recurrence.
    // WHEN: Use when memory is constrained — reduces space from O(m*n) to O(n).

    public int uniquePaths_opt(int m, int n) {
        int[] dp = new int[n];
        Arrays.fill(dp, 1);
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[j] += dp[j - 1];
            }
        }
        return dp[n - 1];
    }

    // APPROACH 3: Combinatorics (Math) | O(min(m,n)) time | O(1) space
    // EXPLAIN: Total paths = C(m+n-2, m-1) — choose how many down-moves from all moves.
    // WHEN: Use for O(1) space; requires careful overflow handling with long arithmetic.

    public int uniquePaths(int m, int n) {
        long result = 1;
        int total = m + n - 2;
        int k = Math.min(m - 1, n - 1);
        for (int i = 1; i <= k; i++) {
            result = result * (total - k + i) / i;
        }
        return (int) result;
    }

    // APPROACH 4: Recursive Memoization | O(m*n) time | O(m*n) space
    // EXPLAIN: paths(i,j) = paths(i-1,j) + paths(i,j-1); cache with 2D memo table.
    // WHEN: Use when top-down recursion is more natural to derive on a whiteboard.

    public int uniquePaths_memo(int m, int n) {
        int[][] memo = new int[m][n];
        for (int[] row : memo) Arrays.fill(row, -1);
        return memoHelper(m - 1, n - 1, memo);
    }

    private int memoHelper(int i, int j, int[][] memo) {
        if (i == 0 || j == 0) return 1;
        if (memo[i][j] != -1) return memo[i][j];
        memo[i][j] = memoHelper(i - 1, j, memo) + memoHelper(i, j - 1, memo);
        return memo[i][j];
    }

    // APPROACH 5: Two-Row DP | O(m*n) time | O(n) space
    // EXPLAIN: Alternate between prev and curr arrays; explicit two-array rolling.
    // WHEN: Use when you prefer explicit separation of prev/curr rows over in-place update.

    public int uniquePaths_twoRow(int m, int n) {
        if (m == 1 || n == 1) return 1;
        int[] prev = new int[n];
        int[] curr = new int[n];
        Arrays.fill(prev, 1);
        for (int i = 1; i < m; i++) {
            curr[0] = 1;
            for (int j = 1; j < n; j++) {
                curr[j] = curr[j - 1] + prev[j];
            }
            int[] tmp = prev; prev = curr; curr = tmp;
        }
        return prev[n - 1];
    }
}

// Made with Bob
